src/app/loadout-builder/process-worker/auto-stat-mod-utils.ts

```diff
@@ -0,0 +1,267 @@
+// the effect is the same.
+const dontSplitTheseAgain: { [statHash in ArmorStatHashes]: ArmorStatHashes[] } = {
+  [StatHashes.Mobility]: [],
+  [StatHashes.Resilience]: [],
+  [StatHashes.Recovery]: [],
+  [StatHashes.Discipline]: [],
+  [StatHashes.Intellect]: [],
+  [StatHashes.Strength]: [],
+};
+
+for (const statHash_ of Object.keys(dontSplitTheseAgain)) {
+  const statHash = Number(statHash_) as ArmorStatHashes;
+  const largeMod = largeStatMods[statHash];
+  const smallMod = minorStatMods[statHash];
+
+  for (const otherStatHash_ of Object.keys(dontSplitTheseAgain)) {
+    const otherStatHash = Number(otherStatHash_) as ArmorStatHashes;
+    const otherLargeMod = largeStatMods[otherStatHash];
+    const otherSmallMod = minorStatMods[otherStatHash];
+    // If splitting otherMod leaves us in the same or a worse position than splitting the current mod
+    if (otherLargeMod.cost <= largeMod.cost && otherSmallMod.cost >= smallMod.cost) {
+      dontSplitTheseAgain[statHash].push(otherStatHash);
+    }
+  }
+}
+
+export interface ModsPick {
+  costs: number[];
+  modHashes: number[];
+}
+
+type ModsWorkingSet = {
+  largeMods: { hash: number; cost: number; statHash: ArmorStatHashes }[];
+  smallMods: { hash: number; cost: number }[];
+};
+
+export interface GeneralModsCache {
+  generalModCosts: number[];
+  statOrder: ArmorStatHashes[];
+  autoStatMods: boolean;
+  cache: { [statsString: string]: ModsPick[] };
+  cacheHits: number;
+  cacheMisses: number;
+  cacheSuccesses: number;
+}
+/**
+ * Okay, so LO should automatically assign stat mods to hit minimum required stats.
+ * Auto mod assignment *must be* correct: If there is a way to pick stat mods and assign
+ * other mods and stat mods so that they fit on the items and the minimum stats are hit,
+ * LO must be able to find it. This is surprisingly tricky: Imagine we need one tier of
+ * strength and one tier of recovery. We could pick the mods in different ways:
+ *
+ * * [large recovery (4), large strength (3)]
+ * * [large recovery (4), small strength (1), small strength (1)]
+ * * [large strength (3), small recovery (2), small recovery (2)]
+ * * [small recovery (2), small recovery (2), small strength (1), small strength (1)]
+ *
+ * Each of these picks is pareto-optimal wrt. leftover energy. There exist armor + mod
+ * picks for which exactly one of these picks fits and no other pick can fit.
+ * This becomes even worse when LO has to assign bucket-independent mods like combat or
+ * activity mods, because there might be a permutation of BI mods that doesn't
+ * leave enough space for any of these picks and another permutation that does.
+ *
+ * What saves us here is that stat mods don't have an element or mod slot requirement, so we
+ * don't have to test all permutations of stat mods. Instead, we just sort the costs of each pick
+ * descending and compare with the leftover energy capacities, also sorted descending.
```

Comment:
I've used this trick in the past when we had a faster mod assignment algorithm. To extend it you can actually sort by energy capacities and by element type (think groups of mods for each element in descending order) and use that to assign mods. While I never verified it's correctness @nev-r and I couldn't break it and I am pretty sure it would have been provable as a greedy algorithm. We had to ditch that when legacy mod slots became a a thing as that conditional handling broke the underlying assumptions allowing the sort to solve the issue.